File: app/src/main/java/com/hault/codex_java/data/local/dao/WorldDao.java
package com.hault.codex_java.data.local.dao;
import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.RawQuery;
import androidx.room.Transaction;
import androidx.room.Update;
import androidx.sqlite.db.SupportSQLiteQuery;
import com.hault.codex_java.data.model.World;
import com.hault.codex_java.data.model.relations.WorldWithDetails;
import java.util.List;
@Dao
public interface WorldDao {
    @Insert
    void insert(World world);

    @Update
    void update(World world);

    @Delete
    void delete(World world);

    @Query("SELECT * FROM worlds ORDER BY name ASC")
    LiveData<List<World>> getAllWorlds();

    @Query("SELECT * FROM worlds WHERE id = :id")
    LiveData<World> getWorldById(int id);

    @RawQuery(observedEntities = World.class)
    LiveData<List<World>> search(SupportSQLiteQuery query);

    @Transaction
    @Query("SELECT * FROM worlds WHERE id = :worldId")
    LiveData<WorldWithDetails> getWorldWithDetails(int worldId);
}

File: app/src/main/java/com/username/codex/data/local/WorldDao.kt
package com.username.codex.data.local

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import com.username.codex.data.model.World
import kotlinx.coroutines.flow.Flow

@Dao
interface WorldDao {

    @Insert
    suspend fun insert(world: World)

    @Update
    suspend fun update(world: World)

    @Delete
    suspend fun delete(world: World)

    @Query("SELECT * FROM worlds ORDER BY name ASC")
    fun getAllWorlds(): Flow<List<World>>
}
